rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,591 @@
+* Struct name prefixes indicates which side of the API is responsible for populating the struct:
+  * `SE_` prefix: filled by TensorFlow.
+  * `SP_` prefix: filled by plugins, except `struct_size` which is also filled by TensorFlow when TensorFlow passes it to a callback.
+
+See proposed C API below:
+
+```cpp
+#include <stddef.h>
+#include <stdint.h>
+
+#define SE_MAJOR 0
+#define SE_MINOR 0
+#define SE_REVISION 1
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef SP_Stream_st* SP_Stream;
+typedef SP_Event_st* SP_Event;
+typedef SP_Timer_st* SP_Timer;
+typedef TF_Status* (*TF_StatusCallbackFn)(void*);
+
+#ifndef TF_BOOL_DEFINED
+#define TF_BOOL unsigned char
+#endif // TF_BOOL_DEFINED
+
+#ifndef TF_OFFSET_OF_END
+#define TF_OFFSET_OF_END(TYPE, MEMBER) (offsetof(TYPE, MEMBER) + sizeof(((TYPE *)0)->MEMBER))
+#endif // TF_OFFSET_OF_END
+
+typedef struct SP_TimerFns {
+  size_t struct_size;
+  void* ext;
+  uint64_t (*nanoseconds)(SE_Timer timer);
+  uint64_t (*microseconds)(SE_Timer timer);
+} SP_TimerFns;
+
+#define SP_TIMER_FNS_STRUCT_SIZE TF_OFFSET_OF_END(SP_TimerFns, microseconds)
+
+typedef struct SP_AllocatorStats {
+  size_t struct_size;
+  void* ext;
+  int64_t num_allocs;
+  int64_t bytes_in_use;
+  int64_t peak_bytes_in_use;
+  int64_t largest_alloc_size;
+
+  int8_t has_bytes_limit;
+  int64_t bytes_limit;
+
+  int64_t bytes_reserved;
+  int64_t peak_bytes_reserved;
+
+  int8_t has_bytes_reservable_limit;
+  int64_t bytes_reservable_limit;
+
+  int64_t largest_free_block_bytes;
+} SP_AllocatorStats;
+
+#define SP_ALLOCATORSTATS_STRUCT_SIZE TF_OFFSET_OF_END(SP_AllocatorStats, largest_free_block_bytes)
+
+typedef enum SE_EventStatus {
+  SE_EVENT_UNKNOWN,
+  SE_EVENT_ERROR,
+  SE_EVENT_PENDING,
+  SE_EVENT_COMPLETE,
+} SE_EventStatus;
+
+typedef struct SE_Options {
+  size_t struct_size;
+  void* ext;
+  int32_t ordinal;
+} SE_Options;
+
+#define SE_OPTIONS_STRUCT_SIZE TF_OFFSET_OF_END(SE_Options, ordinal)
+
+typedef struct SE_DeviceMemoryBase {
+  size_t struct_size;
+  void* ext;
+  void* opaque;
+  uint64_t size;
+  uint64_t payload;
+} SE_DeviceMemoryBase;
+
+#define SE_DEVICE_MEMORY_BASE_STRUCT_SIZE TF_OFFSET_OF_END(SE_DeviceMemoryBase, payload)
+
+typedef struct SP_Device {
+  size_t struct_size;
+  void* ext;
+  const char* name;
+  size_t name_len;
+
+  // Device vendor can store handle to their device representation
+  // here.
+  void* device_handle;
+
+  // Any kind of data that plugin device might want to store.
+  void* ext;
+
+} SP_Device;
+
+#define SP_DEVICE_STRUCT_SIZE TF_OFFSET_OF_END(SP_Device, data)
+
+typedef struct SP_StreamExecutor {
+  size_t struct_size;
+  void* ext;
+
+  /*** ALLOCATION CALLBACKS ***/
+  // Synchronously allocates size bytes on the underlying platform and returns
+  // a DeviceMemoryBase representing that allocation. In the case of failure,
+  // nullptr is returned.
+  // memory_space is reserved for a potential future usage and should be set
+  // to 0.
+  TF_DeviceMemoryBase* (*allocate)(
```

Comment:
Need another allocate API for host memory allocation, counter part of se::HostMemoryAllocate(uint64 size),  this memory is made known to device runtime for effectively copying the data from device and host memory, such as cuMemHostAlloc/cudaMallocHost

Reply:
Added:
```cpp
  // Allocate host memory.
  void* (*host_memory_allocate)(uint64_t size);
  
  // Deallocate host memory.
  void (*host_memory_deallocate)(void *mem);
```

Reply:
seems also need another allocate API which is used for Unified memory allocation, counter part of se::UnifiedMemoryAllocate/se::UnifiedMemoryDeallocate, this seems useful when the memory is overscribed.

Reply:
Sounds good. Does the `unified_memory_allocate`/`unified_memory_deallocate` name still sound good or should we use a different name?

Reply:
I think the name(unified_memory_allocate/unified_memory_deallocate)  is ok

Reply:
and I think se::DeviceDescription may need to add a new flag, like bool support_unified_memory. Current StreamExecutor seems provides cuda compute capability to identify whether the device support unified memory in the tensorflow, which is not suitable for the plugged device.
```
   if (per_process_gpu_memory_fraction > 1.0 ||
       gpu_options.experimental().use_unified_memory()) {
     int cc_major = 0, cc_minor = 0;
     if (!se->GetDeviceDescription().cuda_compute_capability(&cc_major,
                                                             &cc_minor)) {
       return errors::Internal("Failed to get compute capability for device.");
     }
     if (cc_major < 6) {
       return errors::Internal(
           "Unified memory on GPUs with compute capability lower than 6.0 "
           "(pre-Pascal class GPUs) does not support oversubscription.");
     }
   }
```

Reply:
Just to make sure, is unified memory a memory address space that can be accessed by any device?

Unrelated to this specific question, but so far I haven't added DeviceDescription to the API. I wasn't sure if it is useful. Do you think any of the other fields in DeviceDescription would be useful for pluggable devices or just `support_unified_memory`:
https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/device_description.h;l=40?q=DeviceDescription%20streamexecutor

Reply:
yes, unified memory is  a single memory address space accessible from any device(cpu, gpu...) in a system, it is useful when memory is oversubscribed on the device(such as input is too big and device memory is not enough) and user don't want the program fail. Current DeviceDescription seems has no fields which is suitable for other devices. for cuda, it use cuda_compute_capability field in DeviceDescription to check whether the StreamExecutor support unified memory,  I think it may not suitable for other platform.

Reply:
The checked in code here has been updated to include `allocate_unified_memory`, `deallocate_unified_memory` functions. Also, `support_unified_memory` is added to `SP_Platform` (+ some refactoring). 
https://github.com/tensorflow/tensorflow/commit/b6cc708f01ffd7690da74456a230a6cdc64db8d6

I haven't propagated `supports_unified_memory` to `DeviceDescription` yet, probably in next change.